Add `zip` extract support for Windows

huacnlee opened the pull request:
Release Notes:

- [x] Fixed install Node.js runtime and NPM lsp installation on Windows.
- [x] Update Node runtime command to execute on Windows with no window popup.

Ref #9619, #9424

huacnlee commented on crates/fs/src/fs.rs:
```diff
@@ -177,6 +183,34 @@ impl Fs for RealFs {
         Ok(())
     }
 
+    async fn extract_zip_file(
```
This method I have tested, and it work.

huacnlee commented on crates/fs/src/fs.rs:
```diff
@@ -1155,6 +1189,35 @@ impl Fs for FakeFs {
         Ok(())
     }
 
+    async fn extract_zip_file(
```
But I don't know, how to test this FakeFS method.

maxdeviant commented on crates/util/test_data/test.tar.gz:
Let's not put binary files in the repo like this.

This is exactly how the `xz` attack happened.

huacnlee replied:
Can I move this binary data into the test file as a string?

maxdeviant replied:
I think what I'd like to see is for the tests to create the archive and then extract it, that way we don't have any compressed bytes hanging out in the repo.

huacnlee replied:
I have just added `compress_gz`, `compress_tar_gz`, `compress_zip` function for this test.

They may not work with complex case, we need to test in futures and then to improve them. But I think it's ok for currently just for `extract` testing.

huacnlee commented:
<img width="948" alt="image" src="https://github.com/zed-industries/zed/assets/5518/102c67b7-4882-4fa0-b810-540fc7e2cbe2">

maxdeviant commented on crates/util/src/archive.rs:
I don't think this functionality really belongs in `util`.

For one, it makes `util` a much heavier package, as it now has all of the archive-related dependencies in its dependency tree.

I think if we're looking to extract this somewhere else we should add a new `archive` crate.

huacnlee replied:
This have done.

huacnlee commented:
Now, Node.js runtime has fixed and Copilot works.

https://github.com/zed-industries/zed/assets/5518/e27cedf9-6aa1-4a94-9734-89e257f6704e

huacnlee commented:
Node runtime have been passed test.

https://github.com/zed-industries/zed/assets/5518/eed3243f-e201-47db-9504-7d70ba61d013

huacnlee commented:
Hi @maxdeviant, @mikayla-maki 

I think I have done of all about Windows install zip and Node.js runtime install, we can review and testing now.

I have to do a lot of file change, because there are too many issues on Windows. They are mostly issues that have different path on Windows.

huacnlee commented on crates/node_runtime/src/node_runtime.rs:
```diff
@@ -202,6 +217,10 @@ impl NodeRuntime for RealNodeRuntime {
             let mut command = Command::new(node_binary);
             command.env_clear();
             command.env("PATH", env_path);
+            // Set proxy arg, if user's env has `http_proxy`.
```
In China, we have to set http_proxy. Otherwise, the npmjs.org can't connect. 

I added this because the below line have `env_clear`, so add this to keep to use `http_proxy` env.

eth0net commented:
Anything I can do to help with this? Was trying to solve the same issue and found this PR 🙂 Those force-pushes are killing me trying to work out what's going on with the current state of the PR so thought it better I just ask 😅

huacnlee commented on crates/node_runtime/src/node_runtime.rs:
```diff
@@ -215,29 +223,36 @@ impl NodeRuntime for RealNodeRuntime {
             }
 
+            #[cfg(windows)]
+            command.creation_flags(windows::Win32::System::Threading::CREATE_NO_WINDOW.0);
```
@maxdeviant I'd like to improve the `Command::new` in entire of the Zed, there have a lot of place need to do this.

> For example: `crates\languages\src\go.rs`

So, can I add a helper in `util::process:command` in `util` crate to create `smol::process:Command` in shortly with Windows special `creation_flags`? Then other place can just use this method to create command.

> Oh, I found that the comment was in the wrong place. I should post in PR #11156

maxdeviant reviewed (changes requested):
After multiple attempts to review this PR, I think it's touching a few too many parts.

I think we should refocus on just fixing Node runtime support on Windows and leave the other bits untouched.

We can inline the zip extraction into the `node_runtime` crate. We don't need to add a new `archive` crate or touch any of the `tar` or `tar.gz`-related code.

huacnlee commented on crates/extension/src/wasm_host/wit/since_v0_0_6.rs:
```diff
@@ -325,35 +324,18 @@ impl ExtensionImports for WasmState {
                     futures::pin_mut!(body);
-                    self.host.fs.create_file_with(&zip_path, body).await?;
-
-                    let unzip_status = std::process::Command::new("unzip")
```
@maxdeviant 

How about the part? 

I make `extract_zip` method before , because here it requires the new zip extraction method, the `unzip` command is not working on Windows. And node_runtime and more places also need to.

huacnlee commented:
I have updated:

1. Moved `archive` into `node_runtime` crate as a mod, I keep the methods not just only the zip, we many uses that in the future.
2. Reverted other crate changes, just keep the node_runtime.
3. The `since_v0_0_6.rs` still use `unzip` command to extract zip file, this can't work on Windows, I think we can make author PR to discuss this.

huacnlee commented:
I just did a fresh installation of the Node.js runtime on Windows. It seems that we have to modify the root_path part of Copilot (Commit: https://github.com/zed-industries/zed/pull/11156/commits/5131e1a5db1ea941e1f72e253af6c170d545eb19), otherwise it will not start and will directly crash.

huacnlee committed:
Add `zip` unarchive support, and fix node_runtime on Windows platform.

maxdeviant committed:
Merge branch 'main' into pr/11156

maxdeviant committed:
Clean up

maxdeviant committed:
Fix ordering

maxdeviant reviewed (approved):
Thanks, much better!

CharlesChen0823 commented:
> However, the parts that have been merged so far and the panic problem caused by Copilot's root_path unwrap will be fixed. Then the most Node.js related ones can already be run.

this should fixed, currently open immediaterly crashed.
